### PR TITLE
Add new docker image for running kafka and zookeeper in single container

### PR DESCRIPTION
**Pull request by ravisingal**

### CODE REVIEW

File: kafka-zookeeper/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,38 @@
+FROM alpine:3.12 AS kafka
+
+ENV SCALA_VERSION 2.12
+ENV KAFKA_VERSION 2.3.0
+
+WORKDIR /kafka
```
Comment (buchi-busireddy) [inline kafka-zookeeper/Dockerfile:6]: Should we use `/opt/kafka` for consistency?

File: kafka-zookeeper/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+COPY wait-for-zookeeper.sh /kafka/bin
+COPY start.sh /kafka/bin
+
+# Share the same base image to reduce layers
+FROM hypertrace/java:11
+
+MAINTAINER Hypertrace "https://www.hypertrace.org/"
+
+WORKDIR /kafka
```
Comment (buchi-busireddy) [inline kafka-zookeeper/Dockerfile:18]: duplicate line

File: kafka-zookeeper/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+RUN adduser -g '' -h /kafka -D kafka
+
+COPY --from=kafka --chown=kafka /kafka /kafka
+
+USER kafka
```
Comment (buchi-busireddy) [inline kafka-zookeeper/Dockerfile:24]: Declare this above and use $USER in the create command and chown commands.

File: kafka-zookeeper/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,38 @@
+FROM alpine:3.12 AS kafka
```
Comment (buchi-busireddy) [inline kafka-zookeeper/Dockerfile:1]: Curious, why is this needed?

File: kafka-zookeeper/install.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+# Set explicit, basic configuration
+cat > config/server.properties <<-EOF
+broker.id=0
+zookeeper.connect=127.0.0.1:2181
+replica.socket.timeout.ms=1500
+log.dirs=/kafka/logs
+auto.create.topics.enable=true
+offsets.topic.replication.factor=1
+listeners=PLAINTEXT://0.0.0.0:9092,PLAINTEXT_HOST://0.0.0.0:19092
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT
```
Comment (buchi-busireddy) [inline kafka-zookeeper/install.sh:26]: add a comment that advertised listeners will be appended later.

File: kafka-zookeeper/start.sh [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  listeners=PLAINTEXT://:9092
+  # Have internal docker producers and consumers use the normal hostname:9092, and outside docker localhost:19092
+  echo advertised.listeners=PLAINTEXT://${HOSTNAME}:9092,PLAINTEXT_HOST://localhost:19092 >> /kafka/config/server.properties
+else
+  # Have internal docker producers and consumers use the normal hostname:9092, and outside docker the advertised host on port 19092
+  echo "advertised.listeners=PLAINTEXT://${HOSTNAME}:9092,PLAINTEXT_HOST://${KAFKA_ADVERTISED_HOST_NAME}:19092" >> /kafka/config/server.properties
+fi
+
+echo Starting Kafka
+/kafka/bin/kafka-run-class.sh -name kafkaServer -Dlog4j.configuration=file:/kafka/config/log4j.properties kafka.Kafka /kafka/config/server.properties
```
Comment (buchi-busireddy) [inline kafka-zookeeper/start.sh:17]: is the Kafka main class name correct? Just `kafka.Kafka`?

**Comment by buchi-busireddy** [comment]
@ravisingal a high level comment: Did we check if there are any ways to run Kafka and ZK in the same JVM instead of two different processes in same Docker container? That would avoid some failure scenarios where one is running but not the other.

File: kafka-zookeeper/Dockerfile (hunk 4 above)
Comment (ravisingal) [reply kafka-zookeeper/Dockerfile:1]: This is a temp docker container which is used to fetching and extracting kafka tar. Later, we will copy the extracted folder to the final container image. This won't contribute any layer in the final image.

File: kafka-zookeeper/start.sh (hunk 6 above)
Comment (ravisingal) [reply kafka-zookeeper/start.sh:17]: yes.

File: kafka-zookeeper/Dockerfile (hunk 3 above)
Comment (ravisingal) [reply kafka-zookeeper/Dockerfile:24]: you want to set `USER` environment variable?

File: kafka-zookeeper/Dockerfile (hunk 2 above)
Comment (ravisingal) [reply kafka-zookeeper/Dockerfile:18]: This is a separate container.

**Comment by ravisingal** [comment]
> @ravisingal a high level comment: Did we check if there are any ways to run Kafka and ZK in the same JVM instead of two different processes in same Docker container? That would avoid some failure scenarios where one is running but not the other.

I've checked kafka quick start page (https://kafka.apache.org/quickstart). It also starts zookeeper as a separate process.

File: kafka-zookeeper/Dockerfile (hunk 1 above)
Comment (ravisingal) [reply kafka-zookeeper/Dockerfile]: done.

File: kafka-zookeeper/install.sh (hunk 5 above)
Comment (ravisingal) [reply kafka-zookeeper/install.sh:26]: done.

File: kafka-zookeeper/Dockerfile (hunk 4 above)
Comment (buchi-busireddy) [reply kafka-zookeeper/Dockerfile:1]: Got it. thanks

File: kafka-zookeeper/Dockerfile (hunk 3 above)
Comment (ravisingal) [reply kafka-zookeeper/Dockerfile]: done.
